[ENH] BEP030: Functional Near-Infrared Spectroscopy (#802)
bids-standard/bids-specification: src/04-modality-specific-files/11-functional-near-infrared-spectroscopy.md

and not all pairs have to be recorded or represented in the data.  Note that the source
and detector names used in the channel specifications are specified in the `*_optodes.tsv`
file below. The required columns in the `*_channels.tsv` file MUST be ordered as listed below.

The BIDS specification supports several types of fNIRS devices which output raw data in
different forms. The type of measurement is specified in the `type` column. For example,
when measurements are taken with a continuous wave (CW) device that saves the data
as optical density, the `type` should be `NIRSCWOPTICALDENSITY` and the `units` should be `unitless`,
this is equivalent to SNIRF data type `dOD`.

The following columns MUST be present:

{{ MACROS___make_columns_table(
   {
      "name__channels": "REQUIRED",
      "type__nirs_channels": "REQUIRED",
      "source__channels": "REQUIRED",
      "detector__channels": "REQUIRED",
      "wavelength_nominal": "REQUIRED",
      "units__nirs": "REQUIRED",
      "sampling_frequency": "OPTIONAL but REQUIRED if `SamplingFrequency` is to `n/a` in `_nirs.json`",
      "orientation_component": "OPTIONAL but REQUIRED if `type` is `ACCEL`, `GYRO` or `MAGN`",
   }
) }}

> The following columns SHOULD be present:

Reviewer: These are all marked OPTIONAL:

```suggestion
The following columns MAY be present:
```

      "source__optodes": "OPTIONAL",
   }
) }}

Example:
```Text
name    type         x          y         z          template_x    template_y   template_z
A1      source       -0.0707    0.0000    -0.0707    -0.07         0.00         0.07
Fz      detector     0.0000     0.0714    0.0699     0.0           0.07         0.07
S1      source       -0.2707    0.0200    -0.1707    -0.03         0.02         -0.2
D2      detector     0.0022     0.1214    0.0299     0.0           0.12         0.03
VisS2   source       -0.1707    0.1200    -0.3707    -0.1          0.1          -0.4
VisD4   detector     0.0322     0.2214    0.2299     0.02          0.22         0.23
```

## Coordinate System JSON (`*_coordsystem.json`)

{{ MACROS___make_filename_template(datatypes=["nirs"], suffixes=["coordsystem"]) }}

A `*_coordsystem.json` file is used to specify the fiducials, the location of anatomical landmarks,
and the coordinate system and units in which the position of optodes and landmarks is expressed.
Fiducials are objects with a well-defined location used to facilitate the localization of sensors
and co-registration, anatomical landmarks are locations on a research subject such as the nasion

> (for a detailed definition see the coordinate system description in the BIDS specification).

Author: Great catch.

> The `*_coordsystem.json` is REQUIRED if the optional `*_optodes.tsv` is specified. If a corresponding

Reviewer: ```suggestion
The `*_coordsystem.json` is REQUIRED if the optional `*_optodes.tsv` is present. If a corresponding
```

anatomical MRI is available, the locations of anatomical landmarks in that scan should also be stored

> in the `*_T1w.json` file which goes alongside the fNIRS data.

Reviewer: I would need an example here to understand what's being proposed, but this feels like it's going to be extraordinarily difficult to validate. What is a corresponding anatomical MRI? Any T1w image in the subject, or one linked by some metadata? What happens if there are two T1w images? What should the additional metadata be?

Author: As @sappelhoff notes in https://github.com/bids-standard/bids-specification/pull/802#pullrequestreview-1121994111:

> HOWEVER, if this is what's also in MEG, EEG, etc. (which I suspect), then it's fine leaving it as is (for consistency) and tackling it all together in an upcoming PR at some point (doesn't need to be done by NIRS team)

I just checked, and this same text is in EEG. It is NOT in iEEG or MEG.

@sappelhoff I am thinking of just removing this sentence, do you foresee any issues with that?

Reviewer 2: No, please do leave it aligned with the specification for [EEG](https://bids-specification.readthedocs.io/en/stable/04-modality-specific-files/03-electroencephalography.html#coordinate-system-json-_coordsystemjson), as scalp optode localization compares to EEG scalp electrode localization. For MEG (using head coils) and iEEG (using CT imaging data) the localization procedures are different. 
@effigies the corresponding T1w is the one that is pointed to with the `IntendedFor` in the `coordinates.json`

Author: Thanks @robertoostenveld . Marking this as resolved as IntendedFor was added today.

| GYRO                        | Gyrometer channel, one channel for each orientation. An extra column `component` for the axis of the orientation MUST be added to the `*_channels.tsv` file (x, y or z).     |
| MAGN                        | Magnetomenter channel, one channel for each orientation. An extra column `component` for the axis of the orientation MUST be added to the `*_channels.tsv` file (x, y or z). |
| MISC                        | Miscellaneous                                                                                                                                                                |

Example:

```Text
Name         type                   source      detector      wavelength_nominal   units
S1-D1        NIRSCWAMPLITUDE        A1          Fz            760                  V
S1-D1        NIRSCWAMPLITUDE        A1          Fz            850                  V
S1-D2        NIRSCWAMPLITUDE        A1          Cz            760                  V
S2-D1        NIRSCWAMPLITUDE        A2          Fz            760                  V
S3-D4        NIRSCWAMPLITUDE        VisS2       VisD4         760                  V
```

## Optode description (`*_optodes.tsv`)

> {{ MACROS___make_filename_template(datatypes=["nirs"], suffixes=["optodes"]) }}

Reviewer 3: This filename template allows for `*_optodes.json`

We don't have rules for those in the validator. Is it a valid filename or included by mistake? Didn't see any examples of them in the two bids-examples PR.

Reviewer 4: Don't all TSV files need to support sidecar files for defining column metadata? The rules for optodes files say `additional_columns: allowed_if_defined`, so there needs to be some place where those additional columns may be defined.

VisD4   detector     0.0322     0.2214    0.2299     0.02          0.22         0.23
```

## Coordinate System JSON (`*_coordsystem.json`)

{{ MACROS___make_filename_template(datatypes=["nirs"], suffixes=["coordsystem"]) }}

A `*_coordsystem.json` file is used to specify the fiducials, the location of anatomical landmarks,
and the coordinate system and units in which the position of optodes and landmarks is expressed.
Fiducials are objects with a well-defined location used to facilitate the localization of sensors
and co-registration, anatomical landmarks are locations on a research subject such as the nasion
(for a detailed definition see the coordinate system description in the BIDS specification).
The `*_coordsystem.json` is REQUIRED if the optional `*_optodes.tsv` is specified. If a corresponding
anatomical MRI is available, the locations of anatomical landmarks in that scan should also be stored
in the `*_T1w.json` file which goes alongside the fNIRS data.

Not all fNIRS systems provide 3D coordinate information or digitization capabilities.
In this case, only x and y are specified and z is `"n/a"`.

Fields relating to the fNIRS optode positions:

> {{ MACROS___make_sidecar_table(["nirs.CoordinateSystem", "nirs.CoordinateSystemDescriptionRec"]) }}

Reviewer 3: @effigies @tsalo The enumerated types in this table are very long. Should this go back to the old way of doing things? Is there a change that should be made to the macro?

Reviewer 4: I wouldn't be opposed to adding a new parameter like `n_values_to_combine` that would replace the enumerated values with a sentence like "For a list of valid values for this field, see the associated glossary entry." And the "associated glossary entry" text would link to the glossary entry. I can take a crack at that.

Reviewer 4: @rwblair WDYT of #1260?

Reviewer 3: It looked good to me. I looked at some of the collapse styles in mkdocs material, but for the simplicity and pdf rendering I think this is a good option.

# Functional Near-Infrared Spectroscopy

Support for functional Near-Infrared Spectroscopy (fNIRS) was developed as a
[BIDS Extension Proposal](../07-extensions.md#bids-extension-proposals).
Please see [Citing BIDS](../01-introduction.md#citing-bids)
on how to appropriately credit this extension when referring to it in the

> context of the academic literature.

Author: Luca and I will meet in the coming week(s) to discuss. I think it will be 2-3 months away (1 month for second half of fnirs contributors to review, followed by one month for BIDS contributors to review).

Several [example fNIRS datasets](https://github.com/bids-standard/bids-examples#fnirs-datasets)

> have been formatted using this specification and can be used for practical guidance when curating a new dataset.

Author: There is only one in the examples repo. There are more available with full data files listed in the manuscript. One from Donders, 2 from me. I can create more for the bids-examples repo, but the variation will be limited as I simply use MNE-BIDS to create them. What's ideal here?

> have been formatted using this specification and can be used for practical guidance when curating a new dataset.

Reviewer 5: You are right, for the example we strive to have diverse examples -- having many homogeneous ones wouldn't be too helpful. So perhaps Robert's and Helena's example (https://github.com/bids-standard/bids-examples/pull/305#issuecomment-1160090888) would be cool to have.

> have been formatted using this specification and can be used for practical guidance when curating a new dataset.

Reviewer 6: I may be able to add more soon...currently seeking permission from data collectors and authors

> have been formatted using this specification and can be used for practical guidance when curating a new dataset.

Author: There are now two example datasets as PRs. Ideally, we would have more.

## fNIRS recording data

{{ MACROS___make_filename_template(datatypes=["nirs"], suffixes=["nirs", "events", "channels", "optodes", "coordsystem"]) }}

Only the Shared Near Infrared Spectroscopy Format ([SNIRF](https://github.com/fNIRS/snirf))
file specification is supported in BIDS. The SNIRF
specification supports one or more fNIRS datasets to be stored in a single
`.snirf` file. However, to be BIDS compatible, each SNIRF file MUST contain

> only a single run. A limited set of fields from the SNIRF specification are

Author: There is no JS reader for SNIRF. And I know of no initiatives to achieve this.

However, the SNIRF files are just HDF5 files. And I found a javascript library called `jsfive` that claims to be "A pure javascript HDF5 file reader". So I think it would be feasible in principle. 

I am glad this is not a blocker. I am not versed in javascript, so this would be a very large undertaking for me to implement.

> only a single run. A limited set of fields from the SNIRF specification are

Reviewer 5: sounds cool! Would you mind opening an issue on bids-validator that tracks exactly this feature for a future enhancement?

> only a single run. A limited set of fields from the SNIRF specification are

Author: https://github.com/bids-standard/bids-validator/issues/1472

-   Optode - Refers to either a source or detector.

-   Channel - A paired coupling of a source and a detector with one specific wavelength of light.
    It is common for a single Source-Detector pair to result in two or more channels
    with different wavelengths.

### Sidecar JSON (`*_nirs.json`)

It is common within the fNIRS community for researchers to build their own caps
and optode holders to position their sources and detectors, or for optodes to
be directly attached to the scalp with adhesive. To facilitate description of
the wide variety of possible configurations, several fields are RECOMMENDED within
the `*_nirs.json` file.
To clarify the usage and interaction of these fields, the following examples are provided:

-   If a commercial cap such as EasyCap actiCAP 64 Ch Standard-2 was used:
    ```JSON
    "CapManufacturer": "EasyCap",
    "CapManufacturersModelName": "actiCAP 64 Ch Standard-2",
    "NIRSPlacementScheme": "n/a"
    ```

-   If an EasyCap was used but with custom positions,
    as may be done by cutting custom holes in the cap,
    was used:
    ```JSON
    "CapManufacturer": "EasyCap",

>     "CapManufacturersModelName": "custom",

Reviewer 7: Maybe our example on the finger/foot task can help here. See fore example [this file](https://github.com/robertoostenveld/bids-examples/commit/98bb25cd26e67f7de4d2f0770678695b842fe6eb#diff-e8e80f34d6a56741d95def140333e60bfdeb3ba02aa351d6a34506a8043dfe3c). 
We set the CapManufacturersModelName to "headcap with print, customized (L)", but I like your idea to add a field CapModelCustomized. It would then become something like this:
	"CapManufacturer": "Artinis Medical Systems",
	"CapManufacturersModelName": "headcap with print (L)",
        "CapModelCustomized": true,
        
Regarding the NIRSPlacementSchema we solved it this way: 
	"NIRSPlacementScheme": "see optode_layout.pdf: 2 groups over the left and right dlPFC, 2 groups over the left and right PPC, 1 group over the left M1 and PMC",
We then added the optode_layout.pdf in the .bidsignore file. (I misunderstood this first and placed the optode_layout.pdf in .bidsignore folder instead of adding the file in the main folder and writing the filename in the .bidsignore text file, but will update the correct version soon).

>     "CapManufacturersModelName": "custom",

Reviewer 2: I am not really enthusiastic about the boolean `CapModelCustomized` metadata field, nor about `CapManufacturersModelName` having to be selected from a list of restricted keywords. There will be other models in the future and we don't want to update the spec. And writing that it is custom does yet not give sufficient details (e.g. how was it modified?). The proposal is not optimal from human readability point of view, and although it is machine readable, the machine (software) cannot make much sense of it. 

I would prefer something like this 

`CapManufacturersModelName`:  `headcap with print, size L, it was modified by adding holes for the optodes  according to the NIRSPlacementScheme and optode_layout.pdf `

Are there actually well-defined NIRS cap model names given by the manufacturers? For EEG we for example have the [ECI types](https://www.electrocap.nl/index.php?page=caps&cat=1) and  the [Easycap types](https://www.easycap.de/wp-content/uploads/2019/10/Partial_PRODUCT-OVERVIEW-08-2019.pdf) which are detailed and presumably unique codes, not just a "head cap with print, sized L".

>     "CapManufacturersModelName": "custom",

Reviewer 8: No, there are not yet any well defined cap models. We would like to come up with 3ish standards, but this will take a few years to achieve. For now, your proposed solution sounds good and seems to me that it will be future proof to these 3ish standards we hope to achieve.

>     "CapManufacturersModelName": "custom",

Author: Thank you all for the detailed discussion. This has been one of the tougher aspects of the specification to get right, and I think we are almost there. Based on the comments above, how about...

| Key name                  | Requirement Level | Data type                  | Description                                                                                                                                                                                                                                                                                                                                                                                                                  |
|---------------------------|-------------------|----------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CapManufacturer           | RECOMMENDED       | string                     | Name of the cap manufacturer (for example, "EasyCap"). If a custom-made cap is used then the string `custom` MUST be used. If no cap was used, such as with optodes that are directly taped to the scalp, then the string  `none`  MUST be used and the NIRSPlacementScheme field MAY be used to specify the optode placement.                                                                                                   |
| CapManufacturersModelName | RECOMMENDED       | string                     | Manufacturer's designation of the cap model (for example, "actiCAP 64 Ch Standard-2"). If there is no official model number then a description may be provided (for example, Headband with print (S-M)). If a cap from a standard manufacturer was modified, then the field SHOULD be used to describe the cap and modification. If no cap was used, then the CapManufacturer field MUST be none and this field MUST be n/a. |
| NIRSPlacementScheme       | RECOMMENDED       | string or array of strings | Placement scheme of NIRS optodes. Either the name of a standardized placement system (for example, "10-20"), an array of standardized position names (for example, ["Cz", "Pz"]), or description of placement should be used. Reference to further documentation or figures can be placed here. If a standard cap was used, then it should be specified in CapManufacturer and CapManufacturersModelName and this field MAY be used to describe the placement scheme. |

And then we would have the following examples...

-   If a commercial cap such as EasyCap actiCAP 64 Ch Standard-2 was used:
    ```JSON
    "CapManufacturer": "EasyCap",
    "CapManufacturersModelName": "actiCAP 64 Ch Standard-2",
    "NIRSPlacementScheme": "10-20"
    ```

-   If an Artinis Medical Systems cap with custom positions,
    as may be done by cutting custom holes in the cap,
    was used:
    ```JSON
    "CapManufacturer": "Artinis Medical Systems",
    "CapManufacturersModelName": "headcap with print, size L, it was modified by adding holes for the optodes according to the NIRSPlacementScheme and optode_layout.pdf",
    "NIRSPlacementScheme": "see optode_layout.pdf: 2 groups over the left and right dlPFC, 2 groups over the left and right PPC, 1 group over the left M1 and PMC"
    ```

-   If a completely custom cap was knitted:
    ```JSON
    "CapManufacturer": "custom",
    "CapManufacturersModelName": "custom knitted cap with holes for optodes according to the NIRSPlacementScheme and optode_knitted_layout.jpg",
    "NIRSPlacementScheme": "see optode_knitted_layout.jpg: 2 groups over the left and right dlPFC, 2 groups over the left and right PPC"
    ```

-   If no cap was used and optodes were taped to the scalp
    at positions Cz, C1 and C2:
    ```JSON
    "CapManufacturer": "none",
    "CapManufacturersModelName": "n/a",
    "NIRSPlacementScheme": ["Cz", "C1", "C2"],
    ```

> The proposal is not optimal from human readability point of view, and although it is machine readable, the machine (software) cannot make much sense of it.

The proposal is no better at being machine readable, but is it more human readable? Alternative proposals are welcome.

>     "CapManufacturersModelName": "custom",

Reviewer 5: I like this suggestion -- some comments:

- Can we standardize the `optode_layout.pdf` to behave something like `_photo.jpg`? https://bids-specification.readthedocs.io/en/latest/04-modality-specific-files/03-electroencephalography.html#landmark-photos-_photojpg ... or do you all judge that this would overload `_photo` by too much? @robertoostenveld 
- We should explicitly declare `"custom"` and `"none"` as reserved keywords for `CapManufacturer` to be used for the cases outlined above. I wouldn't want to be ambiguous whether people may also use `"n/a"` in those cases.

>     "CapManufacturersModelName": "custom",

Reviewer 2: I don't like for `CapManufacturer` that _"If a custom-made cap is used then the string "custom" MUST be used."_ I would say that the manufacturer is nearly always well defined. It can be the name of a company, the name of a university, the name of a technical support group within a department, or the name of the tailor. E.g. https://www.ru.nl/socialsciences/technicalsupportgroup/ or http://kinh-do.nl would be well-defined places where I might have a custom cap made in Nijmegen, and I might specify "RU socsci TSG"  or "Kinh-Do". Or if it is not an organization, the person doing the sewing or knitting can be specified. The cap being custom made-to-order or being a confection off-the-shelf cap needs to be represented in the `CapManufacturersModelName` field.

So I propose

```
"CapManufacturer": "the mother of the first author",
"CapManufacturersModelName": "custom knitted cap with holes for optodes according to the NIRSPlacementScheme and optode_knitted_layout.jpg",
"NIRSPlacementScheme": "see optode_knitted_layout.jpg: 2 groups over the left and right dlPFC, 2 groups over the left and right PPC"
```

Re `CapManufacturersModelName`: I propose to scrap "standard" in the phrase  "...standard manufacturer...". The manufacturer is not standard, but the cap might be a standard model for the manufacturer (like a 501 for Levis).

>     "CapManufacturersModelName": "custom",

Author: I have address this in ad74ef2b1383901c5d1780c32273c1bd2c864d15 and e678b15379f0d653d262ba9661a9a334374791f7.

This is a subtle but important discussion. Please review my changes and let me know if I missed some details. Thank you all for your attention to detail and input above.

    "NIRSPlacementScheme": "n/a"
    ```

-   If a completely custom cap was knitted:
    ```JSON
    "CapManufacturer": "custom",

>     "CapManufacturersModelName": "custom",

Author: Sure, I will update the text too. I will wait until we have a consensus above on the fields before making the change

>     "CapManufacturersModelName": "custom",

Author: Thanks @sappelhoff I 100% agree with you
I made the changes in e678b15379f0d653d262ba9661a9a334374791f7, please let me know if they don't sufficently address the issue.

    "NIRSPlacementScheme": "n/a"
    ```

-   If no cap was used and optodes were taped to the scalp
    at positions Cz, C1 and C2:
    ```JSON
    "CapManufacturer": "none",
    "CapManufacturersModelName": "n/a",

>     "NIRSPlacementScheme": ["Cz", "C1", "C2"],

Author: I will make it clearer in the text once agreed above.

>     "NIRSPlacementScheme": ["Cz", "C1", "C2"],

Author: Thanks @sappelhoff  I switched to `none`, that was a great suggestion. I decided against glued or taped as different groups do this slightly differently with different terms, so I thought it was going to be too specific.

I made the changes in e678b15379f0d653d262ba9661a9a334374791f7, please let me know if they don't sufficently address the issue.

      "TaskDescription": "RECOMMENDED",
      "Instructions": "RECOMMENDED",
      "CogAtlasID": "RECOMMENDED",
      "CogPOID": "RECOMMENDED",
      "DeviceSerialNumber": "RECOMMENDED",
      "RecordingDuration": "RECOMMENDED",
      "HeadCircumference": "RECOMMENDED",
      "HardwareFilters": "RECOMMENDED",
      "SubjectArtefactDescription": "RECOMMENDED",
   }
) }}

{{ MACROS___make_metadata_table(
   {
      "CapManufacturer": ("RECOMMENDED", "If a custom-made cap is used then the string `custom` should be used. If no cap was used, such as with optodes that are directly taped to the scalp, then the string `none` should be used and the `NIRSPlacementScheme` field may be used to specify the optode placement."),
      "CapManufacturersModelName": ("RECOMMENDED", "If there is no official model number then description may be provided (for example, `Headband with print (S-M)`). If a cap from a standard manufacturer was modified, then the field should be set to `custom`. If no cap was used, then the `CapManafacturer` field should be `none` and this field should be `n/a`."),
   }
) }}

Specific fNIRS fields that MUST be present:

{{ MACROS___make_metadata_table(
   {

>       "SamplingFrequency__nirs": ("REQUIRED", "Sampling frequency (in Hz) of all the data in the recording, regardless of their type (for example, `12`).  If individual channels have different sampling rates, then the field here should be specified as `n/a` and the values  should be specified in the `sampling_frequency` column in channels.tsv."),

Author: Agreed, thanks.

  "SamplingFrequency": 3.7,
  "NIRSChannelCount": 56,
  "NIRSSourceOptodeCount": 16,
  "NIRSDetectorOptodeCount": 16,
  "ACCELChannelCount": 0,
  "SoftwareFilters": "n/a",
  "RecordingDuration": 233.639,
  "HardwareFilters": {"Highpass RC filter": {"Half amplitude cutoff (Hz)": 0.0159, "Roll-off": "6dBOctave"}},
  "CapManafacturer": "NIRx",
  "CapManufacturersModelName": "Headband with print (S-M)",
}
```

### Participant file

The participants.tsv and participants.json files are RECOMMENDED to describe properties of participants such as
age, sex, handedness. If the participants.tsv file exists, it MUST contain the column participant_id
and SHOULD contain the age field, as this is required for calculation of age-specific pathlength factors.
If the dataset contains multiple groups, for example patients and controls, then
a column labeled `group` MAY be included (see [participants.tsv](https://bids-specification.readthedocs.io/en/stable/03-modality-agnostic-files.html#participants-file)).

```text
participant_id	age	 sex	hand  group
sub-01       	34	 M	    n/a   patient
sub-02	        32	 F	    n/a   control

> ```

Reviewer 9: agreed

> ```

Author: Also agreed :)
Actioned in cdbc04f639446bcc93a3b9dafaee904c3174774d

Example:

```Text
Name         type                   source      detector      wavelength_nominal   units
S1-D1        NIRSCWAMPLITUDE        A1          Fz            760                  V
S1-D1        NIRSCWAMPLITUDE        A1          Fz            850                  V
S1-D2        NIRSCWAMPLITUDE        A1          Cz            760                  V
S2-D1        NIRSCWAMPLITUDE        A2          Fz            760                  V
S3-D4        NIRSCWAMPLITUDE        VisS2       VisD4         760                  V
```

## Optode description (`*_optodes.tsv`)

{{ MACROS___make_filename_template(datatypes=["nirs"], suffixes=["optodes"]) }}

File that provides the location and type of optodes. Note that coordinates MUST be
expressed  in Cartesian coordinates according to the NIRSCoordinateSystem and
NIRSCoordinateSystemUnits fields in `*_coordsystem.json`. If an `*_optodes.tsv`
file is specified, a `*_coordsystem.json` file MUST be specified as well.
The order of the required columns in the `*_optodes.tsv` file MUST be as listed below.

The x, y, and z positions are for measured locations, for example, with a polhemus

> digitizer. If you also have idealised positions, where you wish the optodes to be

Reviewer 5: ```suggestion
digitizer. If you also have idealized positions, where you wish the optodes to be
```

> placed, these can be listed in the template values. SNIRF contains arrays for both
> the 3D and 2D locations of data. The BIDS format MUST store the 3D locations if
> available, and only the 2D locations if 3D positions are unavailable. The storage
> of 2D locations would be indicated by the z field containing an `n/a` value.

Reviewer 5: ```suggestion
placed, these can be listed in the template values
(for example for "template positions" computed on a sphere).
SNIRF contains arrays for both
the 3D and 2D locations of data. The BIDS format MUST store the 3D locations if
available, and only the 2D locations if 3D positions are unavailable. The storage
of 2D locations would be indicated by the z field containing an `n/a` value.
```

for a resource on computing template positions a sphere, I have shared: https://stefanappelhoff.com/eeg_positions/ ... but we don't need to add it here, because it's mostly for EEG.

The following columns MUST be present:

{{ MACROS___make_columns_table(
   {
      "name__optodes": "REQUIRED",
      "type__optodes": "REQUIRED",
      "x__optodes": "REQUIRED",
      "y__optodes": "REQUIRED",
      "z__optodes": "REQUIRED",
   }
) }}

The following columns MAY be present:

{{ MACROS___make_columns_table(
   {

>       "x__template": "OPTIONAL but REQUIRED if x is `n/a`",
>       "y__template": "OPTIONAL but REQUIRED if x is `n/a`",
>       "z__template": "OPTIONAL but REQUIRED if x is `n/a`",

Reviewer 5: ```suggestion
      "x__template": "OPTIONAL but REQUIRED if `x` is `n/a`",
      "y__template": "OPTIONAL but REQUIRED if `x` is `n/a`",
      "z__template": "OPTIONAL but REQUIRED if `x` is `n/a`",
```

D2      detector     0.0022     0.1214    0.0299     0.0           0.12         0.03
VisS2   source       -0.1707    0.1200    -0.3707    -0.1          0.1          -0.4
VisD4   detector     0.0322     0.2214    0.2299     0.02          0.22         0.23
```

## Coordinate System JSON (`*_coordsystem.json`)

{{ MACROS___make_filename_template(datatypes=["nirs"], suffixes=["coordsystem"]) }}

A `*_coordsystem.json` file is used to specify the fiducials, the location of anatomical landmarks,
and the coordinate system and units in which the position of optodes and landmarks is expressed.
Fiducials are objects with a well-defined location used to facilitate the localization of sensors
and co-registration, anatomical landmarks are locations on a research subject such as the nasion
(for a detailed definition see the coordinate system description in the BIDS specification).
The `*_coordsystem.json` is REQUIRED if the optional `*_optodes.tsv` is specified. If a corresponding
anatomical MRI is available, the locations of anatomical landmarks in that scan should also be stored
in the `*_T1w.json` file which goes alongside the fNIRS data.

> Not all fNIRS systems provide 3D coordinate information or digitisation capabilities.
> In this case, the only x and y are specified and z is `"n/a"`.

Reviewer 5: ```suggestion
Not all fNIRS systems provide 3D coordinate information or digitization capabilities.
In this case, only x and y are specified and z is `"n/a"`.
```

# Functional Near-Infrared Spectroscopy

Support for functional Near-Infrared Spectroscopy (fNIRS) was developed as a
[BIDS Extension Proposal](../07-extensions.md#bids-extension-proposals).
Please see [Citing BIDS](../01-introduction.md#citing-bids)
on how to appropriately credit this extension when referring to it in the
context of the academic literature.

Several [example fNIRS datasets](https://github.com/bids-standard/bids-examples#fnirs-datasets)
have been formatted using this specification and can be used for practical guidance when curating a new dataset.

## fNIRS recording data

{{ MACROS___make_filename_template(datatypes=["nirs"], suffixes=["nirs", "events", "channels", "optodes", "coordsystem"]) }}

Only the *Shared Near Infrared Spectroscopy Format* ([SNIRF](https://github.com/fNIRS/snirf))

> file specification is supported in BIDS. The SNIRF

Author: My understanding was that all BEPs once merged were BIDS and that there was no fNIRS-BIDS. I am happy to use which ever name is most consistent with other aspects of BIDS.

> file specification is supported in BIDS. The SNIRF

Reviewer 5: > My understanding was that all BEPs once merged were BIDS and that there was no fNIRS-BIDS

That's correct ... we sometimes still talk about EEG-BIDS or MEG-BIDS e.g., in the publications or informally. As this BEP was called NIRS, I'd vote to change this to NIRS-BIDS, rather than with the "f" prefix.

specification supports one or more fNIRS datasets to be stored in a single
`.snirf` file. However, to be BIDS compatible, each SNIRF file MUST contain
only a single run. A limited set of fields from the SNIRF specification are
replicated  in the BIDS specification. This redundancy allows the data to be
easily parsed by humans and machines that do not have a SNIRF reader at hand,
which improves findability and tooling development.

Raw fNIRS data in the native format, if different from SNIRF, can also
be stored in the [`/sourcedata` directory](../02-common-principles.md#source-vs-raw-vs-derived-data)
along with code to convert the data to

> SNIRF in the [`/code` directory](../02-common-principles.md#storage-of-derived-datasets).

Author: Thanks!

S2-D1        NIRSCWAMPLITUDE        A2          Fz            760                  V
S3-D4        NIRSCWAMPLITUDE        VisS2       VisD4         760                  V
```

## Optode description (`*_optodes.tsv`)

{{ MACROS___make_filename_template(datatypes=["nirs"], suffixes=["optodes"]) }}

File that provides the location and type of optodes. Note that coordinates MUST be
expressed  in Cartesian coordinates according to the NIRSCoordinateSystem and
NIRSCoordinateSystemUnits fields in `*_coordsystem.json`. If an `*_optodes.tsv`
file is specified, a `*_coordsystem.json` file MUST be specified as well.
The order of the required columns in the `*_optodes.tsv` file MUST be as listed below.

The x, y, and z positions are for measured locations, for example, with a polhemus
digitizer. If you also have idealized positions, where you wish the optodes to be
placed, these can be listed in the template values
(for example for "template positions" computed on a sphere).
SNIRF contains arrays for both
the 3D and 2D locations of data. The BIDS format MUST store the 3D locations if
available, and only the 2D locations if 3D positions are unavailable. The storage

> of 2D locations would be indicated by the z field containing an `n/a` value.

Author: Thanks

D2      detector     0.0022     0.1214    0.0299     0.0           0.12         0.03
VisS2   source       -0.1707    0.1200    -0.3707    -0.1          0.1          -0.4
VisD4   detector     0.0322     0.2214    0.2299     0.02          0.22         0.23
```

## Coordinate System JSON (`*_coordsystem.json`)

{{ MACROS___make_filename_template(datatypes=["nirs"], suffixes=["coordsystem"]) }}

A `*_coordsystem.json` file is used to specify the fiducials, the location of anatomical landmarks,
and the coordinate system and units in which the position of optodes and landmarks is expressed.
Fiducials are objects with a well-defined location used to facilitate the localization of sensors
and co-registration, anatomical landmarks are locations on a research subject such as the nasion
(for a detailed definition see the coordinate system description in the BIDS specification).
The `*_coordsystem.json` is REQUIRED if the optional `*_optodes.tsv` is specified. If a corresponding
anatomical MRI is available, the locations of anatomical landmarks in that scan should also be stored
in the `*_T1w.json` file which goes alongside the fNIRS data.

> 

Author: Great suggestion, thanks @guiomar. I have now added the IndendFor field

      "SoftwareVersions": "RECOMMENDED",
      "TaskDescription": "RECOMMENDED",
      "Instructions": "RECOMMENDED",
      "CogAtlasID": "RECOMMENDED",
      "CogPOID": "RECOMMENDED",
      "DeviceSerialNumber": "RECOMMENDED",
      "RecordingDuration": "RECOMMENDED",
      "HeadCircumference": "RECOMMENDED",
      "HardwareFilters": "RECOMMENDED",
      "SubjectArtefactDescription": "RECOMMENDED",
   }
) }}

{{ MACROS___make_metadata_table(
   {
      "CapManufacturer": ("RECOMMENDED", "If a custom-made cap is used then the string `custom` should be used. If no cap was used, such as with optodes that are directly taped to the scalp, then the string `none` should be used and the `NIRSPlacementScheme` field may be used to specify the optode placement."),
      "CapManufacturersModelName": ("RECOMMENDED", "If there is no official model number then description may be provided (for example, `Headband with print (S-M)`). If a cap from a standard manufacturer was modified, then the field should be set to `custom`. If no cap was used, then the `CapManafacturer` field should be `none` and this field should be `n/a`."),
   }
) }}

Specific fNIRS fields that MUST be present:

> TODO: In the table below the SamplingFrequency data type should be "number or n/a".

Author: @sappelhoff do you know how I can modify SamplingFrequency to be "number or n/a" for NIRS, but not change the type for the other modalities?

Reviewer 5: This could be tricky schema-wise. In the past we have sometimes created the same schema entry twice with a small hack. for example the metadata `name` has a different meaning when used in a channels.tsv or an electrodes.tsv. See for example: https://github.com/bids-standard/bids-specification/blob/9201b203ffaa72d83b2fa30d1c61f46f089f77de/src/schema/objects/columns.yaml#L202-L219

I think for `SamplingFrequency`, we'd either have to do something similar (please advise, @tsalo  :-) ) ... or we could discuss making `n/a` an admissible option for `SamplingFrequency` throughout the spec.

I would be open to that if users then specify the sampling frequency in the `sampling_frequency` column in their `channels.tsv` file (as you also suggest in your BEP). Not having any specification of sampling frequency should lead to a validator error. -- Having that said, the only problem I see with that route would be implementing support for that in the validator not only for fnirs, but also meg, eeg, and ieeg cc @rwblair

Author: Thanks. I have implemented `SamplingFrequency__nirs` and the doc renders correctly. We can change to another solution if required later.

{{ MACROS___make_metadata_table(
   {
      "SamplingFrequency": ("REQUIRED", "Sampling frequency (in Hz) of all the data in the recording, regardless of their type (for example, `12`).  If individual channels have different sampling rates, then the field here should be specified as `n/a` and the values  should be specified in the `sampling_frequency` column in channels.tsv."),
      "NIRSChannelCount": "REQUIRED",
      "NIRSSourceOptodeCount": "REQUIRED",
      "NIRSDetectorOptodeCount": "REQUIRED",
      "ACCELChannelCount": "RECOMMENDED but REQUIRED if any channel type is ACC",
      "GYROChannelCount": "RECOMMENDED but REQUIRED if any channel type is GYRO",
      "MAGNChannelCount": "RECOMMENDED but REQUIRED if any channel type is MAGN",
   }
) }}

Specific fNIRS fields that SHOULD be present:

> TODO: The SourceType and DetectorType data types should be "string or mixed".

Author: Simillarly @sappelhoff , how can I add `"mixed"` to the data type for the source and detector? I tried editing the metadata.yaml but I couldnt make it work.

Author: Of course! "mixed" is a string, so this is fine as it is. My bad, I was too in the weeds and missed the bigger picture. Thanks @sappelhoff
